### DISCUSSION TITLE
PID controller fan curve

### DISCUSSION THREAD

**Discussion by val3nt33n** [Rem0o/FanControl.Releases#1609] [Ideas]
Have you considered adding a fan curve based on a PID controller that targets a set temperature or temperature range? The idea is to automatically find the optimal fan speed for keeping the temperature in that range.

**Comment by Rem0o** [comment]
Tried it, but the relation between the fan speed and the temperature is too indirect and slow for a PID to be able to do anything. Also, the "target" temperature will change depending on the load of the component, like idle vs load scenarios.   It just doesn't work great in practice.

**Reply by val3nt33n** [reply]
I see. Thanks for giving it a try.

**Reply by Minxster** [reply]
@Rem0o if you don't mind me dropping in on this. I use your wonder software for my main PCs, but not for my servers. Depending on the use-case, PID controlled fans can be a amazing. I run a server that has had all the fans replaced with Noctua (mainly for noise reasons) and using a PID script really keeps things nice and steady. I understand that this might not be "perfect" for normal PCs or high-performance gaming PCs, but I would say PID has it's uses.

If I may, here's a script that quite a few home-made server people use. It is there to be tinkered with a quite a few people do, but out of the box it already does a fantastic (no pun untended LOL).
https://www.truenas.com/community/resources/fan-scripts-for-supermicro-boards-using-pid-logic.24/

So for someone like me, I'm hard pressed to use anything other than PID to find it's own lowest RPM/PMW for a given temperature setpoint, when it comes to servers.

**Reply by Rem0o** [reply]
Might see it work in a server environment, as loads tend to be a slow changing wave in the context of a high volume web server. But you can also simply set a fixed speed on those, close the door, and never look at them again for the next 2-3 years they'll be operating.

**Reply by Minxster** [reply]
It's definitely a slower change in temperature, but not as much as you think when you've already balanced the fan speed vs temperature. Most servers and infrastructure devices (out of scope 🤣) will use fans excessively so they know there won't be an issue. For home labs/servers it can be quite different and trying to chase energy savings and lower noise, leads to headaches with tuning regular fan curves.

PID isn't with out it's issues and it's a complete shift in thought proces: "please use the fans to keep a steady temperature of X". This is far from what we expect with general PCs 😉 There is certainly a large BTU change compared to PCs 😂. But it's a really good when used for the right reasons.

When I used a simple straight line to calculate temp vs fan rpm/PWM (at 30c use 10%, through to 50c use 100%) to control my server's HDD bay, the whole control system would slowly oscillate and just never balance. I had to try and mess with the "line" to try and smooth it out. It was a nightmare 😂 This, and some other factors, lead me to PID.

I get all of this is an edge case 👍😉 It's an interesting topic, certainly when it comes to something other than PCs 🤷‍♂️
